File: SettingsManager.m
//
//  SettingsManager.m
//  monoleap
//
//  Created by Dan Bar-Yaakov on 8/2/15.
//  Copyright (c) 2015 Dan Bar-Yaakov. All rights reserved.
//

#import "SettingsManager.h"

@implementation SettingsManager

NSString *const LEFT_X_CTRL_VALUE = @"leftXCtrlValue";
NSString *const LEFT_Y_CTRL_VALUE = @"leftYCtrlValue";
NSString *const RIGHT_Y_CTRL_VALUE = @"rightYCtrlValue";
NSString *const RIGHT_X_CTRL_VALUE = @"rightXCtrlValue";
NSString *const PITCH_BEND_ENABLED = @"pitchBendEnabled";
NSString *const KEYSWITCH_ENABLED = @"keySwitchEnabled";
NSString *const VELOCITY_ENABLED = @"velocityEnabled";

NSString *const RIGHT_X_CTRL_ENABLED = @"rightXCtrlEnabled";
NSString *const RIGHT_Y_CTRL_ENABLED = @"rightYCtrlEnabled";
NSString *const LEFT_X_CTRL_ENABLED = @"leftXCtrlEnabled";
NSString *const LEFT_Y_CTRL_ENABLED = @"leftYCtrlEnabled";

NSString *const MIDI_OUT_CH = @"midiOutChannel";
NSString *const SYNTH_ENABLED = @"SynthEnabled";

NSString *const THEME_NAME = @"themeName";
NSString *const FINGER_WIDTH = @"fingerWidth";

+ (id)sharedInstance {
    static SettingsManager *sharedMyManager = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedMyManager = [[self alloc] init];
    });
    return sharedMyManager;
}

- (id)init {
    if (self = [super init]) {
        
        NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
        
        _leftXCtrlValue = [defaults objectForKey:LEFT_X_CTRL_VALUE];
        if (_leftXCtrlValue == nil) {
            _leftXCtrlValue = [NSNumber numberWithInt:1];
        }
        
        _leftYCtrlValue = [defaults objectForKey:LEFT_Y_CTRL_VALUE];
        if (_leftYCtrlValue == nil) {
            _leftYCtrlValue = [NSNumber numberWithInt:71];
        }
        
        _rightYCtrlValue = [defaults objectForKey:RIGHT_Y_CTRL_VALUE];
        if (_rightYCtrlValue == nil) {
            _rightYCtrlValue = [NSNumber numberWithInt:74];
        }
        
        _rightXCtrlValue = [defaults objectForKey:RIGHT_X_CTRL_VALUE];
        if (_rightXCtrlValue == nil) {
            _rightXCtrlValue = [NSNumber numberWithInt: 12];
        }
        
        _pitchBendEnabled = [defaults objectForKey:PITCH_BEND_ENABLED];
        if (_pitchBendEnabled == nil) {
	        _pitchBendEnabled = [NSNumber numberWithBool:true];
        }
        
        _keySwitchEnabled = [defaults objectForKey:KEYSWITCH_ENABLED];
        if (_keySwitchEnabled == nil) {
            _keySwitchEnabled = [NSNumber numberWithBool:false];
        }
        
        _velocityEnabled = [defaults objectForKey:VELOCITY_ENABLED];
        if (_velocityEnabled == nil) {
            _velocityEnabled = [NSNumber numberWithBool:false];
        }
        
        _rightXCtrlEnabled = [defaults objectForKey:RIGHT_X_CTRL_ENABLED];
        if (_rightXCtrlEnabled == nil) {
            _rightXCtrlEnabled = [NSNumber numberWithBool:false];
        }
        
        _rightYCtrlEnabled = [defaults objectForKey:RIGHT_Y_CTRL_ENABLED];
        if (_rightYCtrlEnabled == nil) {
            _rightYCtrlEnabled = [NSNumber numberWithBool:true];
        }
        
        _leftXCtrlEnabled = [defaults objectForKey:LEFT_X_CTRL_ENABLED];
        if (_leftXCtrlEnabled == nil) {
            _leftXCtrlEnabled = [NSNumber numberWithBool:true];
        }
        
        _leftYCtrlEnabled = [defaults objectForKey:LEFT_Y_CTRL_ENABLED];
        if (_leftYCtrlEnabled == nil) {
            _leftYCtrlEnabled = [NSNumber numberWithBool:true];
        }

        _themeName = [defaults objectForKey:THEME_NAME];
        if (_themeName == nil) {
            _themeName = @"Basic";
        }
        
  
        _midiOutChannel = [defaults objectForKey:MIDI_OUT_CH];
        if (_midiOutChannel == nil) {
            _midiOutChannel = [NSNumber numberWithInteger:1];
        }
        
        _synthEnabled = [defaults objectForKey:SYNTH_ENABLED];
        if (_synthEnabled == nil) {
            _synthEnabled = [NSNumber numberWithBool: true];
        }
        
        _fingerWidth = [defaults objectForKey:FINGER_WIDTH];
        if (_fingerWidth == nil) {
            _fingerWidth = [NSNumber numberWithFloat:136];
        }
        
    }
    return self;
}

- (void)dealloc {

}


@end

File: xophone/SettingsManager.swift
//
//  SettingsManager.swift
//  monoleap
//
//  Created by Omer Elimelech on 16/09/2021.
//  Copyright © 2021 Dan Bar-Yaakov. All rights reserved.
//

import Foundation

class SettingsManager: NSObject {
    
    static let sharedInstance = SettingsManager()
    
    var themeName: String? {
        return DefaultsManager.themeName
    }
    var leftXCtrlValue: Int? {
        return DefaultsManager.leftXCtrlValue
    }
    var leftYCtrlValue: Int? {
        return DefaultsManager.leftYCtrlValue
    }
    var rightYCtrlValue: Int? {
        return DefaultsManager.rightYCtrlValue
    }
    var rightXCtrlValue: Int? {
        return DefaultsManager.rightXCtrlValue
    }
    var leftXCtrlEnabled: Bool? {
        return DefaultsManager.leftXCtrlEnabled
    }
    var leftYCtrlEnabled: Bool? {
        return DefaultsManager.leftYCtrlEnabled
    }
    var rightXCtrlEnabled: Bool? {
        return DefaultsManager.rightXCtrlEnabled
    }
    var rightYCtrlEnabled: Bool? {
        return DefaultsManager.rightYCtrlEnabled
    }
    var pitchBendEnabled: Bool? {
        return DefaultsManager.pitchBendEnabled
    }
    var keySwitchEnabled: Bool? {
        return DefaultsManager.keySwitchEnabled
    }
    var velocityEnabled: Bool? {
        return DefaultsManager.velocityEnabled
    }
    var midiOutChannel: String? {
        return DefaultsManager.midiOutChannel
    }
    var synthEnabled: Bool? {
        return DefaultsManager.synthEnabled
    }
    var fingerWidth: Float? {
        return DefaultsManager.fingerWidth
    }
}

class DefaultsManager: NSObject {
    static let defaults = UserDefaults.standard
    struct Constants {
        static let leftXCtrlValue = "leftXCtrlValue"
        static let leftYCtrlValue = "leftYCtrlValue"
        static let rightYCtrlValue = "rightYCtrlValue"
        static let rightXCtrlValue = "rightXCtrlValue"
        static let pitchBendEnabled = "pitchBendEnabled"
        static let keySwitchEnabled = "keySwitchEnabled"
        static let velocityEnabled = "velocityEnabled"
        
        
        static let rightXCtrlEnabled = "rightXCtrlEnabled"
        static let rightYCtrlEnabled = "rightYCtrlEnabled"
        static let leftXCtrlEnabled = "leftXCtrlEnabled"
        static let leftYCtrlEnabled = "leftYCtrlEnabled"
        
        static let midiOutChannel = "midiOutChannel"
        static let synthEnabled = "SynthEnabled"
        
        static let themeName = "themeName"
        static let fingerWidth = "fingerWidth"
    }
    
    static var leftXCtrlValue: Int {
        get { return defaults.value(forKey: Constants.leftXCtrlValue) as? Int ?? 1 }
        set { defaults.setValue(newValue, forKey: Constants.leftXCtrlValue)}
    }
    
    static var leftYCtrlValue: Int {
        get { return defaults.value(forKey: Constants.leftYCtrlValue) as? Int ?? 71 }
        set { defaults.setValue(newValue, forKey: Constants.leftYCtrlValue)}
    }
    
    static var rightYCtrlValue: Int {
        get { return defaults.value(forKey: Constants.rightYCtrlValue) as? Int ?? 74 }
        set { defaults.setValue(newValue, forKey: Constants.rightYCtrlValue)}
    }
    
    static var rightXCtrlValue: Int {
        get { return defaults.value(forKey: Constants.rightXCtrlValue) as? Int ?? 12 }
        set { defaults.setValue(newValue, forKey: Constants.rightXCtrlValue)}
    }
    
    static var pitchBendEnabled: Bool {
        get { return defaults.value(forKey: Constants.pitchBendEnabled) as? Bool ?? true }
        set { defaults.setValue(newValue, forKey: Constants.pitchBendEnabled)}
    }
    
    static var keySwitchEnabled: Bool {
        get { return defaults.value(forKey: Constants.keySwitchEnabled) as? Bool ?? false }
        set { defaults.setValue(newValue, forKey: Constants.keySwitchEnabled)}
    }
    
    static var velocityEnabled: Bool {
        get { return defaults.value(forKey: Constants.velocityEnabled) as? Bool ?? false }
        set { defaults.setValue(newValue, forKey: Constants.velocityEnabled)}
    }

    static var rightXCtrlEnabled: Bool {
        get { return defaults.value(forKey: Constants.rightXCtrlEnabled) as? Bool ?? false }
        set { defaults.setValue(newValue, forKey: Constants.rightXCtrlEnabled)}
    }
    
    static var rightYCtrlEnabled: Bool {
        get { return defaults.value(forKey: Constants.rightYCtrlEnabled) as? Bool ?? true }
        set { defaults.setValue(newValue, forKey: Constants.rightYCtrlEnabled)}
    }
    
    static var leftXCtrlEnabled: Bool {
        get { return defaults.value(forKey: Constants.leftXCtrlEnabled) as? Bool ?? true }
        set { defaults.setValue(newValue, forKey: Constants.leftXCtrlEnabled)}
    }
    
    static var leftYCtrlEnabled: Bool {
        get { return defaults.value(forKey: Constants.leftYCtrlEnabled) as? Bool ?? true }
        set { defaults.setValue(newValue, forKey: Constants.leftYCtrlEnabled)}
    }
    
    static var midiOutChannel: String {
        get { return defaults.value(forKey: Constants.midiOutChannel) as? String ?? "1" }
        set { defaults.setValue(newValue, forKey: Constants.midiOutChannel)}
    }
    
    static var synthEnabled: Bool {
        get { return defaults.value(forKey: Constants.synthEnabled) as? Bool ?? false }
        set { defaults.setValue(newValue, forKey: Constants.synthEnabled)}
    }
    
    static var themeName: String {
        get { return defaults.value(forKey: Constants.themeName) as? String ?? "Basic" }
        set { defaults.setValue(newValue, forKey: Constants.themeName)}
    }
    
    static var fingerWidth: Float {
        get { return defaults.value(forKey: Constants.fingerWidth) as? Float ?? 136.0 }
        set { defaults.setValue(newValue, forKey: Constants.fingerWidth)}
    }

}
